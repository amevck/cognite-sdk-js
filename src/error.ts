// Copyright 2019 Cognite AS

import { AxiosError } from 'axios';

export class CogniteError extends Error {
  public status: number;
  public requestId?: string;
  constructor(errorMessage: string, status: number, requestId?: string) {
    let message = `${errorMessage} | code: ${status}`;
    if (requestId) {
      message += ` | X-Request-ID: ${requestId}`;
    }
    super(message);
    this.status = status;
    this.requestId = requestId;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }
}

/** @hidden */
export function handleErrorResponse(err: AxiosError) {
  let code;
  let message;
  let requestId;
  try {
    code = err.response!.status;
    message = err.response!.data.error.message;
    requestId = (err.response!.headers || {})['X-Request-Id'];
  } catch (_) {
    throw err;
  }
  throw new CogniteError(message, code, requestId);
}
