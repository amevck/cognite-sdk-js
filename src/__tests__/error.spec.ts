// Copyright 2019 Cognite AS

import { AxiosError } from 'axios';
import { CogniteError, handleErrorResponse } from '../error';
import { createErrorReponse } from './testUtils';

const internalIdObject = { id: 4190022127342195 };
const externalIdObject = { externalId: 'abc' };
const event = {
  externalId: 'string',
  startTime: 0,
  endTime: 0,
  description: 'string',
  source: 'string',
};

describe('CogniteError', () => {
  test('without requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status);
    expect(error.status).toBe(status);
    expect(error.requestId).toBeUndefined();
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot(`"Abc | code: 500"`);
  });

  test('with requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const requestId = 'def';
    const error = new CogniteError(errorMessage, status, requestId);
    expect(error.status).toBe(status);
    expect(error.requestId).toBe(requestId);
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot(
      `"Abc | code: 500 | X-Request-ID: def"`
    );
  });

  test('extra field', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status, undefined, {
      missing: [internalIdObject, externalIdObject],
      duplicated: [event],
    });
    expect(() => {
      throw error;
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('handleErrorResponse', () => {
  test('without requestId', () => {
    const axiosError = {
      response: {
        status: 500,
        data: createErrorReponse(500, 'abc'),
      },
    } as AxiosError;

    expect(() => {
      handleErrorResponse(axiosError);
    }).toThrowErrorMatchingInlineSnapshot(`"abc | code: 500"`);
  });

  test('with requestId', () => {
    const axiosError = {
      response: {
        status: 500,
        data: createErrorReponse(500, 'abc'),
        headers: {
          'X-Request-Id': 'def',
        },
      },
    } as AxiosError;

    expect(() => {
      handleErrorResponse(axiosError);
    }).toThrowErrorMatchingInlineSnapshot(
      `"abc | code: 500 | X-Request-ID: def"`
    );
  });

  test('extra fields', () => {
    const status = 500;
    const message = 'abc';
    const xRequestId = 'def';
    const axiosError = {
      response: {
        status,
        data: createErrorReponse(status, message, {
          missing: [internalIdObject, externalIdObject],
          duplicated: [event],
        }),
        headers: {
          'X-Request-Id': xRequestId,
        },
      },
    } as AxiosError;

    expect(() => {
      handleErrorResponse(axiosError);
    }).toThrowErrorMatchingSnapshot();

    try {
      handleErrorResponse(axiosError);
    } catch (e) {
      expect(e.status).toBe(status);
      expect(e.requestId).toBe(xRequestId);
      expect(e.missing).toEqual([internalIdObject, externalIdObject]);
      expect(e.duplicated).toEqual([event]);
    }
  });
});
