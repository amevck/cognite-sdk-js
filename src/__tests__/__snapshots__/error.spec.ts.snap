// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CogniteError extra field 1`] = `
"Abc | code: 500
Missing:
  {\\"id\\":4190022127342195}
  {\\"externalId\\":\\"abc\\"}
Duplicated:
  {\\"externalId\\":\\"string\\",\\"startTime\\":0,\\"endTime\\":0,\\"description\\":\\"string\\",\\"source\\":\\"string\\"}"
`;

exports[`handleErrorResponse extra fields 1`] = `
"abc | code: 500 | X-Request-ID: def
Missing:
  {\\"id\\":4190022127342195}
  {\\"externalId\\":\\"abc\\"}
Duplicated:
  {\\"externalId\\":\\"string\\",\\"startTime\\":0,\\"endTime\\":0,\\"description\\":\\"string\\",\\"source\\":\\"string\\"}"
`;
